### DISCUSSION TITLE
Cross-platform client

### DISCUSSION THREAD

**Discussion by erlend-sh** [elk-zone/elk#3030] [Ideas]
This isn’t so much a request but rather just an open tracker of progress towards an open cross-platform client.

With the rise of Bluesky, many of us are now dividing our microblogging time evenly between both Mastodon and Bluesky, hopping between two separate client experiences, even though the content format is 95% the same.

The prospect of a cross-platform client has been lauded and pursued by open web advocates for a long time:

- https://publicinfrastructure.org/2022/11/09/gobo-2-0-all-your-social-media-in-one-place/
- https://blog.erlend.sh/juicy-clients

There’s at least one live example already, but it’s a closed client: https://openvibe.social/

One possible stepping stone would be a fork of Elk to make a Bluesky client:
https://bsky.app/profile/patak.dev/post/3l72elk6ip32r

Meanwhile, there are several Nuxt/Vue devs now working on standalone Bluesky clients, so even though there’s no fork (which could chart a path to a merger) yet, we can glean from these what a hybrid Masto/Bsky codebase in Elk might look like in the future:

- https://github.com/prazdevs/niji by @prazdevs 
- https://github.com/mimonelu/klearsky @mimonelu
- https://github.com/JuanjoSalvador/sunset-bsky @JuanjoSalvador

**Comment by JuanjoSalvador** [comment]
One of my ideas for the future in Sunset, was adding Mastodon support due to components being pretty reusable, just need to add a different service for Mastodon.

To me, having one client for both would be really amazing and useful. I can help on this project, if it's needed.

**Reply by erlend-sh** [reply]
> just need to add a different service for Mastodon.

Can’t it be done the same way but here then, in the very mature Elk project?

Simplistically put, all I want as a user is two separate app-interfaces (albeit with loads of shared components) that I can toggle between with an account switcher.

**Reply by JuanjoSalvador** [reply]
> Can’t it be done the same way but here then, in the very mature Elk project?

Not sure in Elk, since I didn't check enough the code, but probably.

**Reply by erlend-sh** [reply]
If you’re able to start chipping away at this from the Elk-end then you’ll have the benefit of reviews & feedback from the Elk devs ✨ 

In which case it would instead be a matter of making *Bluesky components* for Elk, yes?

**Comment by theAdhocracy** [comment]
Just adding support. I'm seeing a lot more people either moving to BlueSky or opening an additional account there, but I miss Elk's UI. It would be great if I could use one app on my phone to keep up with both feeds; even better if they could somehow be combined!

EDIT: Returning to say that you _can_ use BlueSky today with Elk via the Sky Bridge server. Seems to work pretty well so far: [Sky Bridge](https://skybridge.fly.dev/)

**Comment by erlend-sh** [comment]
Turns out the aforementioned Gobo is open source, so that provides a very useful reference for a cross-platform client/backend:

https://github.com/iDPI-Umass/gobo-client
https://github.com/iDPI-Umass/gobo-backend
by @freeformflow

I wonder how the approach of Gobo compares to the API calls conversion of [SkyBridge](https://github.com/videah/SkyBridge) by @videah 🤔

**Comment by nileshtrivedi** [comment]
+1 for this.

There is a lot of value in unifying the data model and API for Mastodon/ActivityPub and Bluesky. Microblogging, at least for reading/consuming, is pretty much same across all these platforms, and a unified interface will allow future innovation. There is a place for dedicated libraries like masto.js or atproto sdk, but ideally end-user apps should depend on a higher abstraction. For example, if Elk was built on a unified feed SDK, adding support for Bluesky would have been trivial.

**Comment by erlend-sh** [comment]
Latest news from the Elk/Vue gang:

Bluesky API client:
https://github.com/tsky-dev/tsky

Bluesky web client:
https://github.com/nimbus-town/nimbus

This development aligns with my previously mentioned ‘dual-client solution’:

> Simplistically put, all I want as a user is two separate app-interfaces (albeit with loads of shared components) that I can toggle between with an account switcher.

An alternative to a fully hybrid client could be a dual-client experience, facilitated by a client-switcher (could just be an extension of account switching) on both ends.

Imagine you’re logged into both an Elk and Nimbus client, which already look almost identical. By wiring these clients together, I.e. making them aware of each other, I could quickly switch between Mastodon/Bluesky viewing as needed, even automatically when following a link pointing to the opposite platform.

It may even be possible to bundle both clients together into one app package using Tauri?
